scripts/source_stats.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/env python3
+
+"""Script to compute statistics about source code."""
+
+from __future__ import annotations
+
+__copyright__ = """
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""
+
+import argparse
+from alive_progress import alive_bar  # type:ignore
+from multiprocessing import Pool
+import re
+import termplotlib as tpl  # type:ignore
+from pathlib import Path
+from typing import Dict, List, Optional
+from dataclasses import dataclass, field, asdict
+
+BLANK_RE = re.compile(r"\s*")
+COMMENT_RE = re.compile(r"\s*///*\s*")
+LINE_RE = re.compile(
+    r"""
+    (?P<class_intro>\b(class|struct)\s+(?P<class_name>\w+)\b)|
+    (?P<end_open_curly>{\s*(?P<open_curly_trailing_comment>//.*)?)|
+    (?P<trailing_comment>//.*)|
+    (?P<id>\b\w+\b)
+""",
+    re.X,
+)
+
+
+@dataclass
+class _Stats:
+    """Stats collected while scanning source files"""
+
+    lines: int = 0
+    blank_lines: int = 0
+    comment_lines: int = 0
+    empty_comment_lines: int = 0
+    comment_line_widths: Dict[int, int] = field(default_factory=lambda: {})
+    lines_with_trailing_comments: int = 0
+    classes: int = 0
+    identifiers: int = 0
+    identifier_widths: Dict[int, int] = field(default_factory=lambda: {})
+    ids_per_line: Dict[int, int] = field(default_factory=lambda: {})
+
+    def add_comment_line_width(self, width: int) -> None:
+        if width in self.comment_line_widths:
+            self.comment_line_widths[width] += 1
+        else:
+            self.comment_line_widths[width] = 1
+
+    def add_identifier_width(self, width: int) -> None:
+        if width in self.identifier_widths:
+            self.identifier_widths[width] += 1
+        else:
+            self.identifier_widths[width] = 1
+
+    def add_ids_per_line(self, num_ids: int) -> None:
+        if num_ids in self.ids_per_line:
+            self.ids_per_line[num_ids] += 1
+        else:
+            self.ids_per_line[num_ids] = 1
```

Comment:
Usual way to do this is using https://docs.python.org/3/library/stdtypes.html#dict.get , https://docs.python.org/3/library/collections.html#collections.Counter , or `setdefault` like you do below.

Reply:
Ah, `Counter` is perfect. I somehow didn't find that when trying to figure out how to do this in Python, and ended up making a really bad version of `Counter` here, so thanks.